addons/portal/models/ir_http.py

```diff
@@ -14,6 +19,61 @@
 # Language tools
 # ------------------------------------------------------------
 
+def url_localized(lang_code=None, url=None, canonical_domain=None, prefetch_langs=False, force_default_lang=False):
+    """ Returns the given URL adapted for the given lang, meaning that:
+    1. It will have the lang suffixed to it
+    2. The model converter parts will be translated
+
+    If it is not possible to rebuild a path, use the current one instead.
+    `url_quote_plus` is applied on the returned path.
+
+    It will also force the canonical domain is requested.
+    Eg:
+    - `_get_url_localized(lang_fr, '/shop/my-phone-14')` will return
+        `/fr/shop/mon-telephone-14`
+    - `_get_url_localized(lang_fr, '/shop/my-phone-14', True)` will return
+        `<base_url>/fr/shop/mon-telephone-14`
+    """
+    if not lang_code:
+        lang = request.lang
+    else:
+        lang = request.env['res.lang']._lang_get(lang_code)
+
+    if not url:
+        url = request.httprequest.path + '?' + keep_query()
+
+    # '/shop/furn-0269-chaise-de-bureau-noire-17?' to
+    # '/shop/furn-0269-chaise-de-bureau-noire-17', otherwise -> 404
+    url, sep, qs = url.partition('?')
+
+    # TODO: check that it's taking care of url_rewrite too through werkzeug match
+
+    try:
+        # Re-match the controller where the request path routes.
+        rule, args = request.env['ir.http']._match(url)
+        for key, val in list(args.items()):
+            if isinstance(val, models.BaseModel):
+                if isinstance(val._uid, RequestUID):
+                    args[key] = val = val.with_user(request.uid)
+                if val.env.context.get('lang') != lang.code:
+                    args[key] = val = val.with_context(lang=lang._get_cached('code'))
+                if prefetch_langs:
+                    args[key] = val = val.with_context(prefetch_langs=True)
+        router = http.root.get_db_router(request.db).bind('')
+        path = router.build(rule.endpoint, args)
+    except (NotFound, AccessError, MissingError):
+        # The build method returns a quoted URL so convert in this case for consistency.
+        path = werkzeug.urls.url_quote_plus(url, safe='/')
+    if force_default_lang or lang != request.env['ir.http']._get_default_lang():
+        path = f'/{lang._get_cached("url_code")}{path if path != "/" else ""}'
+
+    if canonical_domain:
+        # canonical URLs should not have qs
+        return werkzeug.urls.url_join(canonical_domain, path)
+
+    return path + sep + qs
```

Comment:
if not url, your add yourself the ? before the keep_query()
so final url will end with ? whatever if qs... what is not expected